packages/firestore/src/util/async_queue.ts

```diff
@@ -279,6 +316,41 @@ export class AsyncQueue {
     return this.enqueueInternal(op);
   }
 
+  enqueueRetryable(op: () => Promise<void>): void {
+    this.verifyNotFailed();
+
+    if (this._isShuttingDown) {
+      return;
+    }
+
+    if (this.firstRetryableOperation) {
+      // If there is already a retryable operation, enqueue the current
+      // operation right after. We want to successfully run the first retryable
+      // operation before running all others.
+      // eslint-disable-next-line @typescript-eslint/no-floating-promises
+      this.firstRetryableOperation.then(() => {
+        this.enqueueRetryable(op);
+      });
+    } else {
+      const deferred = new Deferred<void>();
+      this.firstRetryableOperation = deferred.promise;
+
+      const retryingOp = async (): Promise<void> => {
+        try {
+          await op();
+          deferred.resolve();
+          this.backoff.reset();
+          this.firstRetryableOperation = undefined;
```

Comment:
This chain from the first retryable operation creates a weird circumstance where we could have a train of retryable operations where we wake up, succeed in performing some operation, fail in some later one, and lose the rest of the chain.

It seems like maintaining an explicit queue of these (with a head and a tail) would make this more resistant to arbitrary hiccups like this.

Or maybe I'm not understand what this code is doing exactly, in which case please add comments about how this is supposed to work.